### DISCUSSION TITLE
HTML5 Build Progress

### DISCUSSION THREAD

**Discussion by pahaze** [pahaze/Mistful-Crimson-Morning#4] [Announcements]
Originally asked by @loniop, HTML5 builds were not possible, however, I've started working on the mod again as I'm in a better spot than before. HTML5 now compiles and runs songs without a lot of features. Some events (namely camera ones) have been ported over to hardcoded events so far. I'll show off more as I get there, but the next push will not be until I complete the migration to hardcoded files. Currently, here's [Cannibalism playing on HTML5](https://youtu.be/1sIaLPE8bqY) missing only the stage! I won't answer some questions for a few hours after posting this as it's 5AM for me, but I'll answer anything I can as soon as possible! Once this is finished, anyone can host it of course, but it'll also go up on my website for people who just want to play and not host it on their own.

UPDATE: THE WEB PORT IS FINISHED AND PUSHED!!! You can now build for web!

**Comment by pahaze** [comment]
Cannibalism now plays fully excluding the end animation for SpongeBob! Not sure why that doesn't work, but I'll figure it out. 
![cannibalism near full on html5](https://user-images.githubusercontent.com/69173292/203655933-ba46020d-9084-4ed1-bc80-75dd0aad74c9.png)

**Reply by pahaze** [reply]
Nevermind, fixed it! Turns out the Change Character event is broken with character names (bf, gf, dad) on HTML5 for some reason, but changing it to 0/1/2 works fine :joy: 
![Cannibalism fully working on HTML5](https://user-images.githubusercontent.com/69173292/203658131-ddd8b66a-f8d5-4a6b-ae34-f0dd8d9845e6.png)

**Comment by pahaze** [comment]
Another update... Dehydrated is now added as well as the new UI! Also, Humiliation was added AND shaders were fixed to work on HTML5 and Linux!!!

![Dehydrated (Phase 1) on HTML5](https://user-images.githubusercontent.com/69173292/203839379-7f9a673f-a845-4921-bf07-9a9c6d5fcd11.png)

![Dehydrated (Phase 2) on HTML5](https://user-images.githubusercontent.com/69173292/203841234-5243c685-61c9-4ba8-8f40-7a68551160f6.png)

![Dehydrated (Ending) on HTML5](https://user-images.githubusercontent.com/69173292/203841356-457656cc-6cf0-420c-98ef-483ace4bebab.png)

![Humiliation + Shaders on HTML5](https://user-images.githubusercontent.com/69173292/203841484-d302d67a-03de-402a-9ca8-041f86b8d972.png)

![Humiliation + Shaders on Linux](https://user-images.githubusercontent.com/69173292/203844600-e298c34d-d558-4301-be59-8cc816d9014a.png)

**Reply by THESTORMZ** [reply]
i noticed the shader was missing in linux build, i kinda added my own static like shader after it was built lol

**Reply by pahaze** [reply]
Nice! This is the exact same shader as Windows and I'll be pushing soon. I only have Satisfaction and Doomsday left to hardcode!

**Reply by THESTORMZ** [reply]
Nice, Does this mean I will be able to compile to Html5?

**Reply by pahaze** [reply]
Yup! It'll be able to be compiled and put anywhere.

**Reply by THESTORMZ** [reply]
great! i will actually have a reason to have my web server up lol.

**Reply by pahaze** [reply]
Nice lol

**Reply by THESTORMZ** [reply]
off topic question, what Distro of linux are you using?

**Reply by pahaze** [reply]
Ubuntu 22.04 LTS

**Reply by THESTORMZ** [reply]
ok i was having trouble with that version so i stayed on 20.04.5 LTS

**Reply by pahaze** [reply]
It's worked fine for me for months now. What issues were you having?

**Reply by THESTORMZ** [reply]
problems mostly with Wine, i actually used Wine  to run MCM before i knew how to compile fnf games.

**Reply by pahaze** [reply]
Oh, that's a GLSL thing. With the next push that'll be fixed as Linux and web shared the same issue.

**Reply by THESTORMZ** [reply]
i also have to use Ubuntu 20 to compile certain things like sm64 or other things bc it has compiling errors on Ubuntu 22.

**Reply by pahaze** [reply]
Weird. I don't have that. Are you sure you set up development things right?

**Reply by THESTORMZ** [reply]
yea its the same install as when i was compiling before, just now its been updated to 22.

**Reply by THESTORMZ** [reply]
that made it stop working.

**Reply by pahaze** [reply]
Huh. I wonder why

**Reply by THESTORMZ** [reply]
it only happens when i try compiling sm64 for the web, and a few fnf mods won't compile (usually kade engine)

**Reply by pahaze** [reply]
Probably its Lua or video integration? I dunno. Never tried compiling Kade Engine.

**Reply by THESTORMZ** [reply]
yea

**Comment by pahaze** [comment]
Almost every v1 song is now playable! I plan to add v2 songs and even pre-v1 songs too.

[https://youtu.be/2890NM4Gutg](https://youtu.be/2890NM4Gutg)

**Reply by THESTORMZ** [reply]
v2 songs?

**Reply by pahaze** [reply]
The scrapped V2 has some new songs

**Reply by THESTORMZ** [reply]
oh ok.

**Reply by pahaze** [reply]
Or well, I guess I should say "updated", not "new"

**Reply by THESTORMZ** [reply]
ok

**Reply by THESTORMZ** [reply]
this is one of my favorite mods, im just glad i will soon be able to play it more often.

**Comment by THESTORMZ** [comment]
cool!

**Comment by THESTORMZ** [comment]
what about the arrows on dead hope?

**Reply by pahaze** [reply]
They work on HTML5. Check the newest video. [https://youtu.be/2890NM4Gutg](https://youtu.be/2890NM4Gutg)

**Reply by THESTORMZ** [reply]
there was an html version on kbh games i think it was just a torn apart version of psych engine with the levels badly ported, this is much better it look the same as the computer version.

**Reply by pahaze** [reply]
My version is the exact same as the PC version as I have hardcoded stages and Lua events to Haxe code that works on web.

**Reply by THESTORMZ** [reply]
probably will have to turn off the shaders on my potato chromebook tho lol

**Reply by pahaze** [reply]
Option's there and works perfectly. I've tested everything, Satisfaction and Doomsday are the only songs left before the move to hardcoded stages and events is done.

**Reply by pahaze** [reply]
Turning on Low Quality also disables Phase 2 of Dead Hope and Dehydrated.

**Reply by THESTORMZ** [reply]
oh good to know

**Reply by THESTORMZ** [reply]
and also for the large sprite sheets, how large are they? in file size i mean.

**Reply by pahaze** [reply]
Depends. Dehydrated's phase 2 BG was like 60MB last time I checked ☠️

**Reply by THESTORMZ** [reply]
so im probably going to have to host MCM on my home server and not github if i choose to host it.

**Reply by pahaze** [reply]
Totally up to you. Speeds won't really be any different, all depends on the connection to the server and the player's specs.

**Reply by THESTORMZ** [reply]
yea but github pages deployment time will be pretty long if files are that big.

**Reply by pahaze** [reply]
Oh :thinking: 
Didn't know that

**Reply by THESTORMZ** [reply]
yea there already pretty long to deploy the base fnf game

**Reply by pahaze** [reply]
Ah. I've only tried with my engine (which was a long time ago) and it didn't take too long so I figured it wasn't that bad.

**Reply by THESTORMZ** [reply]
there's just a lot more usage of ghp so it slower.

**Reply by pahaze** [reply]
Makes sense :shrug:

**Comment by THESTORMZ** [comment]
🦑

**Reply by pahaze** [reply]
You miiiiiiiiiight wanna delete this comment. It contains your school's name in it. I don't know if that was intentional or not, but I figured I would let you know in case it wasn't intentional. Also, I'll try to get the new edits pushed by Saturday. I've got some school work I have to catch up on myself. Sorry if the progress seems slow :sweat_smile:

**Reply by pahaze** [reply]
Nevermind, turns out I can edit the message too? :joy:
Anyways, yeah. Hopefully progress within next weekend.

**Reply by THESTORMZ** [reply]
> You miiiiiiiiiight wanna delete this comment. It contains your school's name in it. I don't know if that was intentional or not, but I figured I would let you know in case it wasn't intentional. Also, I'll try to get the new edits pushed by Saturday. I've got some school work I have to catch up on myself. Sorry if the progress seems slow 😅

thanks lol, didn't know that replying by email does that.

**Comment by THESTORMZ** [comment]
hows the progress?

**Reply by pahaze** [reply]
2 more stages to finished 💪

**Comment by pahaze** [comment]
Sorry for not being finished yet, I've been having a weird time, but I'm finally going to finish this tonight or tomorrow. However, afterwards, I'm finished with the mod. I wanted to add V2 songs, but I don't really think I can. Overall, I just want to say thank you all for the support through these bug fixes and web port and being patient!

**Comment by pahaze** [comment]
Satisfaction is now ported! Last song... Doomsday...
![Satisfaction (Phase 1) on HTML5](https://user-images.githubusercontent.com/69173292/211178859-910ade11-4be3-4233-ab8c-44ce877c40b3.png)
![Satisfaction (Phase 2) on HTML5](https://user-images.githubusercontent.com/69173292/211178947-e63d96c0-2d8f-4730-be3a-80eb1813f96b.png)

**Comment by pahaze** [comment]
I feel like there is no way Doomsday should need this many sprites but this is the same amount as the Lua code... Goodbye RAM
![blah](https://user-images.githubusercontent.com/69173292/211199004-3d0b8a56-58ec-4ebb-854e-f06f27bf6ce5.png)

**Comment by pahaze** [comment]
DOOMSDAY IS PORTED!!! It requires ~3GB of RAM to be free due to all the sprites, but the other phases can be disabled in the options! I'll be cleaning up some things and then the web port will be FULLY ready! If anybody hosts this, PLEASE do not credit me. If you want to credit me, just link to the source code.
![Doomsday on HTML5 (Start)](https://user-images.githubusercontent.com/69173292/211207821-513e2f9b-875d-403e-b590-e08e638ba656.png)
![Doomsday on HTML5 (Stage 1)](https://user-images.githubusercontent.com/69173292/211207682-8bd07daa-fbdc-4d0c-a08a-6bcbb9efd65b.png)
![Doomsday on HTML5 (Fire)](https://user-images.githubusercontent.com/69173292/211207939-6e7dab13-d582-4ff0-b1fa-97bd6239676d.png)
![Doomsday on HTML5 (Stage 2)](https://user-images.githubusercontent.com/69173292/211207981-617b4dd7-05cf-43b9-aa89-5a869ed07b9e.png)
![Doomsday on HTML5 (Post Stage 2)](https://user-images.githubusercontent.com/69173292/211208011-49711080-ca1c-4573-9bc6-422221472d1f.png)

**Comment by pahaze** [comment]
WEB PORT IS PUSHED!!! HAVE FUN!!! If there's any issues, come back here and tell me!

**Reply by THESTORMZ** [reply]
thanks!

**Comment by THESTORMZ** [comment]
it says it could not generate *.woff text files at all

**Reply by pahaze** [reply]
That's normal, idk why Haxe does that. It'll play perfectly fine tho!

**Reply by THESTORMZ** [reply]
it actually crashes my computer when i try to even load the settings menu.

**Reply by pahaze** [reply]
Weird. What are you playing on?

**Comment by THESTORMZ** [comment]
when i look at the logs on my server it looks like it tries to load every asset

**Reply by pahaze** [reply]
Ugh, of course it's doing that... Explains the large RAM usage. It never shown that for me though, but at the same time I only debugged with Lime.

**Reply by THESTORMZ** [reply]
it fills all four gigs of my potato chromebook and chromebooks don't crash a tab when that happens the whole thing crashes

**Reply by pahaze** [reply]
I'll look into it but I don't know what I can test on as most of my devices have 8GB min
May be able to pull an old potato laptop out and get it running again

**Reply by THESTORMZ** [reply]
okay, thanks!

**Reply by pahaze** [reply]
No problem! What I'll likely do is seperate stages' files with libraries and put the minimum in each and make secondary ones for large files

**Comment by THESTORMZ** [comment]
is there any new progress in the web build?

**Reply by THESTORMZ** [reply]
?

**Reply by pahaze** [reply]
Sorry, I've not had time to look at it. I just started going back to school in person and I have some personal things going on outside of school. I don't know when I'll be able to look at it.

**Reply by THESTORMZ** [reply]
ah, all good.
